add FormattedMessage component

Frizi opened the pull request:
Added the implementation of FormattedMessage, solves #5 

Use like
```html
<FormattedMessage
  id="welcome"
  defaultMessage="Hello {name}, you have {count, number} {count, plural, one {message} other {messages}}"
  :values="{count: 15}">
  <strong slot="name">Tim</strong>
</FormattedMessage>
```

Generates
> Hello **Tim**, you have 15 messages

It's tested locally, but without unit tests as there is no dependency on Vue yet. Didn't wanted to change the testing config.

rtibbles commented on src/FormattedMessage.js:
```diff
@@ -0,0 +1,39 @@
+        tagName: {type: String, default: 'span'}
+    },
+
+    render (h, ctx) {
```
Just for readability, might be nice to use `createElement` instead of the rather anonymous `h`.

Frizi replied:
It's possible, but I think using `h` is pretty common standard.

rtibbles commented on src/FormattedMessage.js:
```diff
@@ -0,0 +1,39 @@
+    props: {
+        id: {type: String, required: true},
+        defaultMessage: String,
+        values: Object,
```
Wondering if we should require explicitly passing the values into the component, or better to let the parent component's props and data drive this?

Frizi replied:
If you will drive that by props, how are you going to spread the object in? Vue have problems with that.

rtibbles replied:
This is what I mean: https://vuejs.org/v2/guide/components.html#Compilation-Scope

By default when you use a component inside a parent component, all of the interpolated properties should be the parent properties, and not the child properties, so passing in the values into the child component as the values object to then have 'count' referenceable in what is still the parent scope goes against this.

However, this is also parallel to the React Intl component definitions, and I can't see a better way to do it without getting confused between the formatted message syntax and Vue interpolation.

Frizi replied:
It's not about scopes. I'm actually passing the object. In the demo example, it's an inline object, but often it will be something from data or generated from other sources. There is no point in using any form of bindings to do that. I'd say that slots are a stretch already, but there is no real way around that.

rtibbles commented on src/FormattedMessage.js:
```diff
@@ -0,0 +1,39 @@
+            return h(ctx.props.tagName, format(ctx, ctx.props.values))
+        }
+
+        const values = Object.assign({}, ctx.props.values)
```
So here, we would reference the parent - not sure how much more complex this would become, but it obeys the Vue tendency to have variable names resolved in the parent component's scope when it is inline in the template.

Frizi replied:
I'm not sure I understand. How would the usage of this component look like? Not sure about the tendency your are referring to. Vue always references local component's "scope", not parent's. In case of functional, there is no instance, so the short-lived config is passed instead.

rtibbles replied:
The local component scope when these are used is actually the parent's scope (because everything is done inline in the parent template). Can't think of a better way to do this now, so we'll stay parallel with React Intl implementation.

rtibbles commented on src/FormattedMessage.js:
```diff
@@ -0,0 +1,39 @@
+            pos = match.index + match[0].length
+        }
+        children.push(message.substring(pos))
+        return h(ctx.props.tagName, children)
```
What is the `children` argument doing there? From the Vue docs https://vuejs.org/v2/guide/render-function#createElement-Arguments it seems like an array of children should be the third argument to the createElement function?

Frizi replied:
Every single of them except for the name is optional. Passing children as a second one is allowed because of that.

rtibbles commented:
Thanks for responding to comments, I'll do some testing of this component later in the week, merge, and release to npm.

Frizi commented:
Introduced a small change, added `ctx.data` to the `createElement` invocations, so you can pass down data to anything that actually gets rendered. This adds possibility to handle things like `@click` or `class`.

Frizi commented:
@rtibbles Ping.

If you need any support on that, please share what should be done to complete that PR.

rtibbles commented:
Hi Pawel, sorry for not getting back on this, I am in a double crunch at
the moment, but should be able to get this released by the end of next
week. Thanks for your patience!

On Mon, 13 Feb 2017, 15:13 Paweł Grabarz, <notifications@github.com> wrote:

> @rtibbles <https://github.com/rtibbles> Ping.
>
> If you need any support on that, please share what should be done to
> complete that PR.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/learningequality/vue-intl/pull/9#issuecomment-279554678>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABmkvTH1Q98GQ3blSrj0LEUqLKvYptD5ks5rcOOZgaJpZM4LqbId>
> .
>

Frizi commented:
@rtibbles Any updates on that?

rtibbles commented:
Will generate a new release later today.
